src/openms_gui/source/VISUAL/DIALOGS/ToolsDialog.cpp

```diff
@@ -86,36 +95,40 @@ namespace OpenMS
 
     auto label = new QLabel("TOPP tool:");
     main_grid->addWidget(label, 1, 0);
-    QStringList list;
 
-    // Determine all available tools compatible with the layer_type
+    // Determine all available tools compatible with the layer_type_
     tool_map_ = {
             {FileTypes::Type::MZML, LayerData::DataType::DT_PEAK},
             {FileTypes::Type::MZXML, LayerData::DataType::DT_PEAK},
             {FileTypes::Type::FEATUREXML, LayerData::DataType::DT_FEATURE},
             {FileTypes::Type::CONSENSUSXML, LayerData::DataType::DT_CONSENSUS},
             {FileTypes::Type::IDXML, LayerData::DataType::DT_IDENT}
     };
-    // Get a map of all tools
-    const auto& tools = OpenMS::ToolHandler::getTOPPToolList();
-    for (const auto& tool : tools)
+    // Get a map of all tools and utils
+    const auto& tools = ToolHandler::getTOPPToolList();
+    const auto& utils = ToolHandler::getUtilList();
+    // Extract names of all tools and utils
+    QList<String> keys;
+    for (const auto& pair : tools)
     {
-      const String& tool_name = tool.first;
-      Param p = getParamFromIni_(tool_name);
-      std::vector<LayerData::DataType> tool_types = getTypesFromParam_(p);
-      // Check if tool is compatible with the layer type
-      if (std::find(tool_types.begin(), tool_types.end(), layer_type) != tool_types.end())
-      {
-        list << tool_name.toQString();
-      }
+      keys.push_back(pair.first);
+    }
+    for (const auto& pair : utils)
+    {
+      keys.push_back(pair.first);
     }
+    // Check whether tools/utils are compatible with current layer in background and add all compatible tools/utils
+    // to the tools_combo_ when ready (addEntries_)
+    connect(&param_watcher_, &QFutureWatcher<Param>::finished, this, &ToolsDialog::addEntries_);
+    // Warnings shall be send to TOPPView's LogWindow i.e. the given LogWindow (log_)
+    connect(this, &ToolsDialog::warning, this, &ToolsDialog::logWarning_);
+    param_future_ = QtConcurrent::mapped(keys, std::bind(&ToolsDialog::getParamFromIni_, this, std::placeholders::_1));
```

Comment:
not sure if the dependency on `QtConcurrent` is worth the trouble here.
@jpfeuffer does this have implications on packaging for e.g. our installers, conda, KNIME ... etc? It certainly has for the developer installation I imagine.

To avoid this, I think it's possible to do this using std::async, even though the callback is not as convenient e.g.
```
std::vector<std::future<Param>> result;
for (const QString& key : keys)
{
	result.push_back( std::move(std::async(std::launch::async, []() { return getParamFromIni_(key); });) );
}

for (auto& r : results)
{
	while (r.wait_for(std::chrono::milliseconds(10) != std::future_status::ready))
	{ // a bit of a crutch... (a callback would be better, but its a lot more code to get right)
		QCoreApplication::processEvents();
	}
}
// all params are available now ... evaluate them...
```

Reply:
Hmm yes. We kind of automated the dependency deduction and copying of the DLLs but it is not very robust so it would need to be tested immediately after the merge. But if there is a solution that avoids using another dependency for one small part of the code, I would be in favor of that.

Reply:
Hmm does async have a fixed threadpool or just spawn >hundred. Is this an issue? @cbielow

Reply:
good point.
MSVC uses a threadpool, if i remember correctly.
And there is no build-in way to limit the number of threads.
Most linuxes will spawn several thousand. MSVC spawns twice as many as it detects CPU-cores (last time I tested).
So you might actually block the GUI thread...
The only save way to do the things here, would be to use std::hardware_concurrency and spawn only a limited number the async's, each getting assigned its share of TOPP Tools (more than one). The workload is fairly balanced I would imagine, since querying an INI should be dominated by the Tool starting up.

We might also get rid of this approach entirely, if we make sure that INI files for all TOPP tools are statically stored somewhere in OpenMS/share.. for example. This would need a test, which makes sure they are all up-to-date though... and that would run rather slow, but at least, we could read INI files directly. Just an idea...
How long does it take at the moment to populate the list?

Reply:
depends on hdd type. Maybe 1-5s. The idea was to make this code multi-threaded and then later move it to the start of TOPPView so parsing is done during startup and should be finished when the user opens the tool dialog.

Reply:
We used QConcurrency to have Qt handle all threads in a global thread-pool and we don't mix any multi-threading mechanisms.